File: aoc2024/src/python/day7.py
from operator import add, mul
import time
from typing import Set
from aoc_utils import  runIt, PuzzleInput, splitInts
import functools
import math
import re
from itertools import combinations, permutations, product
import sys


testInput = r"""190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20"""

input = PuzzleInput('input/day7.txt', testInput)

lines = input.getInputLines(test=False)
calibrations = list(map(lambda x: x.split(': '), lines))
# print(calibrations)

def checkValue(value, equation, ops) -> bool:
    acc = equation[0]
    for i, op in enumerate(ops):
        acc = op(acc, equation[i+1])
        if (acc > value): return False
    return (acc == value)

def concat(v1:int, v2:int) -> int:
    return int(str(v1) + str(v2))

def part1():
    values = []
    for valueStr, equationStr in calibrations:
        value = int(valueStr)
        equation = splitInts(equationStr, ' ')
        slots = len(equation) - 1
        for opCombo in product([add, mul], repeat=slots):
            if checkValue(value, equation, opCombo):
                values.append(value)
                break
        
    print(f'sum {sum(values)}')

def part2():
    values = []
    for valueStr, equationStr in calibrations:
        value = int(valueStr)
        equation = splitInts(equationStr, ' ')
        slots = len(equation) - 1
        for opCombo in product([add, mul, concat], repeat=slots):
            if checkValue(value, equation, opCombo):
                values.append(value)
                break
        
    print(f'sum {sum(values)}')

runIt(part1, part2)

# my part2 takes almost 10 sec to run, see more efficient version here from norvig:
# def can_be_calibrated(equation: ints, operators=(operator.add, operator.mul)) -> bool:
#     """Can the equation be balanced using '+' and '*' operators?"""
#     target, first, *rest = equation
#     results = [first] # A list of all possible results of the partial computation
#     for y in rest:
#         results = [op(x, y) for x in results if x <= target for op in operators]
#     return target in results

File: aoc2024/src/ts/day7.ts
import { time, timeEnd, timeLog } from "console";
import { readInput } from "./utils/file-utils";
import { combinations } from "./utils/utils";

const lines = readInput("day7", false);

function evaluate(numbers: number[], ops: string[]): bigint {
  let acc = BigInt(numbers[0]);
  for (let i = 1; i < numbers.length; i++) {
    const op = ops[i - 1];
    const num = BigInt(numbers[i]);
    if (op === "+") {
      acc = acc + num;
    } else if (op === "*") {
      acc = acc * num;
    } else if (op === "||") {
      const s = `${acc}${num}`;
      acc = BigInt(s);
    }
  }
  return acc;
}

function part1() {
  let sum = BigInt(0);
  lines.forEach((line) => {
    const [testValue, numbersString] = line.split(":").map((e) => e.trim());
    const numbers = numbersString.split(" ").map((e) => parseInt(e.trim()));
    const opPermutations = combinations(["+", "*"], numbers.length - 1);
    console.log(`${testValue} from ${numbers} - ${opPermutations}`);

    for (let i = 0; i < opPermutations.length; i++) {
      const result = evaluate(numbers, opPermutations[i]);
      if (result === BigInt(testValue)) {
        console.log(`success with ${opPermutations[i]}`);
        sum = sum + result;
        break;
      }
    }
  });

  console.log(`sum: ${sum}`);
}

function part2() {
  let sum = BigInt(0);
  lines.forEach((line) => {
    const [testValue, numbersString] = line.split(":").map((e) => e.trim());
    const numbers = numbersString.split(" ").map((e) => parseInt(e.trim()));
    const opPermutations = combinations(["+", "*", "||"], numbers.length - 1);
    // console.log(`${testValue} from ${numbers} - ${opPermutations.join(" ")}`);

    for (let i = 0; i < opPermutations.length; i++) {
      const result = evaluate(numbers, opPermutations[i]);
      if (result === BigInt(testValue)) {
        // console.log(`success with ${opPermutations[i]}`);
        sum = sum + result;
        break;
      }
    }
  });

  console.log(`sum: ${sum}`);
}

time();
part2();
timeEnd();
